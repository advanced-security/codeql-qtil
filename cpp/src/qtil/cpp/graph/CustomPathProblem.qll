private import qtil.locations.CustomPathProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path problems.
import PathProblem<Location, CppLocatableConfig>
