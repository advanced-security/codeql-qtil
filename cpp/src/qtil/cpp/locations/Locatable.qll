private import qtil.locations.Locatable
private import cpp as cpp

/**
 * A module to declare `Locatable`s specific to C++ for use in other qtil modules.
 */
module CppLocatableConfig implements LocatableConfig<cpp::Location> {
  class Locatable = cpp::Locatable;
}
