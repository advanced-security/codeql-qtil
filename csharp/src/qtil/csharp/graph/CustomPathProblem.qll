private import qtil.locations.CustomPathProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path problems.
import PathProblem<Location, CsharpLocatableConfig>
