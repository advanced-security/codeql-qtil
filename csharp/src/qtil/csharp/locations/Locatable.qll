private import qtil.locations.Locatable
private import csharp as csharp

/**
 * A module to declare `Locatable`s specific to C# for use in other qtil modules.
 */
module CsharpLocatableConfig implements LocatableConfig<csharp::Location> {
  class Locatable = csharp::Element;
}
