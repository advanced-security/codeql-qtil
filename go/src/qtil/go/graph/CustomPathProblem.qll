private import qtil.locations.CustomPathProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path problems.
import PathProblem<DbLocation, GoLocatableConfig>
