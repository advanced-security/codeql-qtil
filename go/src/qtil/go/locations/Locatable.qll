private import qtil.locations.Locatable
private import go as go

/**
 * A module to declare `Locatable`s specific to Go for use in other qtil modules.
 */
module GoLocatableConfig implements LocatableConfig<go::DbLocation> {
  class Locatable = go::Locatable;
}
