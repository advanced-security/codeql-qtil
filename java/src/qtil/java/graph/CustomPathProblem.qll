private import qtil.locations.CustomPathProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path problems.
import PathProblem<Location, JavaLocatableConfig>
