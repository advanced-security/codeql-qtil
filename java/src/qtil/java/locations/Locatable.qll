private import qtil.locations.Locatable
private import java

/**
 * A module to declare `Locatable`s specific to Java for use in other qtil modules.
 */
module JavaLocatableConfig implements LocatableConfig<Location> {
  class Locatable = Element;
}
