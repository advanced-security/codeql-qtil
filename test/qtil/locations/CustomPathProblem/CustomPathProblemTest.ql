/**
 * @name Custom Path Problem Example
 * @description This example demonstrates how to define a custom path problem in C++ using Qtil. It
 *   identifies paths from top-level variables to constructors that are called during their
 *   initialization.
 * @id qtil-example-custom-path-problem
 * @severity info
 * @kind path-problem
 */

import cpp
import cpp as cpp
import qtil.locations.Locatable
import qtil.locations.CustomPathProblem
import CustomPathProblemCpp

/** Defines cpp location behavior; this will be moved to qtil.cpp eventually. */
module CustomPathProblemCpp {
  module ElementConfig implements LocatableConfig<Location> {
    class Locatable = cpp::Locatable;
  }

  import PathProblem<Location, ElementConfig>
}

/**
 * Defines a custom path problem configuration for identifying paths from top-level variables to
 * constructors that are called during their initialization.
 */
module CallGraphPathProblemConfig implements CustomPathProblemConfigSig {
  /**
   * Since we are tracking flow from variable initialization to constructor calls, that means the
   * nodes in our path problem will be variables (roots), function calls (edges), and constructors
   * (end nodes).
   */
  class Node extends Locatable {
    Node() {
      this instanceof Function or this.(Variable).isTopLevel() or this instanceof FunctionCall
    }
  }

  /** Start searching from variable nodes */
  predicate start(Node n) { n instanceof Variable }

  /** If we reach a constructor, we have identified "problematic" flow from a variable */
  predicate end(Node n) {
    exists(Function f, Class c |
      n = f and
      c.getAConstructor() = f
    )
  }

  predicate edge(Node a, Node b) {
    // Add an edge from variables to the function calls in that variable's initializer.
    exists(Variable var, Expr initializer, FunctionCall fc |
      var.getInitializer().getExpr() = initializer and
      fc.getParent*() = initializer and
      a = var and
      b = fc
    )
    or
    // Supposing we have reached a function call to some function `mid()`, then the next step in
    // the path problem will be one of the function calls in `mid()`.
    exists(FunctionCall fc, Function mid, FunctionCall next |
      mid = fc.getTarget() and
      next.getEnclosingFunction() = mid and
      a = fc and
      b = next
    )
    or
    // Add an edge from function calls to constructors, which are the end nodes.
    exists(FunctionCall fc, Function endFunc |
      fc.getTarget() = endFunc and
      end(endFunc) and
      a = fc and
      b = endFunc
    )
  }
}

// Import the custom path problem configuration and define the problem.
//
// This automaticall generates the `nodes` and `edges` predicates based on the configuration that
// make the path traceable for users.
import CustomPathProblem<CallGraphPathProblemConfig>

from Variable var, Function ctor
where problem(var, ctor) // This finds for paths from variables to constructors
select var, var, ctor, "Initialization of variable $@ calls constructor $@", var, var.getName(),
  ctor, ctor.getName()
