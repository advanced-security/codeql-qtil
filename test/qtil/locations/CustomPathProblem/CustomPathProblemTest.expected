edges
| test.cpp:17:8:17:9 | call to C1 | test.cpp:2:13:2:14 | C1 |  |  |
| test.cpp:22:12:22:13 | call to f1 | test.cpp:17:8:17:9 | call to C1 |  |  |
| test.cpp:33:5:33:6 | g1 | test.cpp:33:10:33:11 | call to f1 |  |  |
| test.cpp:33:10:33:11 | call to f1 | test.cpp:17:8:17:9 | call to C1 |  |  |
| test.cpp:34:5:34:6 | g2 | test.cpp:34:10:34:11 | call to f2 |  |  |
| test.cpp:34:10:34:11 | call to f2 | test.cpp:22:12:22:13 | call to f1 |  |  |
| test.cpp:37:4:37:5 | c1 | test.cpp:37:4:37:5 | call to C1 |  |  |
| test.cpp:37:4:37:5 | call to C1 | test.cpp:2:13:2:14 | C1 |  |  |
| test.cpp:38:5:38:6 | g5 | test.cpp:38:10:38:13 | call to C3 |  |  |
| test.cpp:38:10:38:13 | call to C3 | test.cpp:10:13:10:14 | C3 |  |  |
| test.cpp:39:5:39:6 | g6 | test.cpp:39:17:39:18 | call to f2 |  |  |
| test.cpp:39:17:39:18 | call to f2 | test.cpp:22:12:22:13 | call to f1 |  |  |
nodes
| test.cpp:2:13:2:14 | C1 | semmle.label | C1 |
| test.cpp:10:13:10:14 | C3 | semmle.label | C3 |
| test.cpp:17:8:17:9 | call to C1 | semmle.label | call to C1 |
| test.cpp:22:12:22:13 | call to f1 | semmle.label | call to f1 |
| test.cpp:33:5:33:6 | g1 | semmle.label | g1 |
| test.cpp:33:10:33:11 | call to f1 | semmle.label | call to f1 |
| test.cpp:34:5:34:6 | g2 | semmle.label | g2 |
| test.cpp:34:10:34:11 | call to f2 | semmle.label | call to f2 |
| test.cpp:37:4:37:5 | c1 | semmle.label | c1 |
| test.cpp:37:4:37:5 | call to C1 | semmle.label | call to C1 |
| test.cpp:38:5:38:6 | g5 | semmle.label | g5 |
| test.cpp:38:10:38:13 | call to C3 | semmle.label | call to C3 |
| test.cpp:39:5:39:6 | g6 | semmle.label | g6 |
| test.cpp:39:17:39:18 | call to f2 | semmle.label | call to f2 |
#select
| test.cpp:33:5:33:6 | g1 | test.cpp:33:5:33:6 | g1 | test.cpp:2:13:2:14 | C1 | Initialization of variable $@ calls constructor $@ | test.cpp:33:5:33:6 | g1 | g1 | test.cpp:2:13:2:14 | C1 | C1 |
| test.cpp:34:5:34:6 | g2 | test.cpp:34:5:34:6 | g2 | test.cpp:2:13:2:14 | C1 | Initialization of variable $@ calls constructor $@ | test.cpp:34:5:34:6 | g2 | g2 | test.cpp:2:13:2:14 | C1 | C1 |
| test.cpp:37:4:37:5 | c1 | test.cpp:37:4:37:5 | c1 | test.cpp:2:13:2:14 | C1 | Initialization of variable $@ calls constructor $@ | test.cpp:37:4:37:5 | c1 | c1 | test.cpp:2:13:2:14 | C1 | C1 |
| test.cpp:38:5:38:6 | g5 | test.cpp:38:5:38:6 | g5 | test.cpp:10:13:10:14 | C3 | Initialization of variable $@ calls constructor $@ | test.cpp:38:5:38:6 | g5 | g5 | test.cpp:10:13:10:14 | C3 | C3 |
| test.cpp:39:5:39:6 | g6 | test.cpp:39:5:39:6 | g6 | test.cpp:2:13:2:14 | C1 | Initialization of variable $@ calls constructor $@ | test.cpp:39:5:39:6 | g6 | g6 | test.cpp:2:13:2:14 | C1 | C1 |
